src/bvh/optimization.rs

```diff
@@ -0,0 +1,457 @@
+// shapes, but also their new AABBs into optimize().
+
+impl BVH {
+    /// Optimizes the `BVH` by batch-reorganizing updated nodes.
+    /// Based on https://github.com/jeske/SimpleScene/blob/master/SimpleScene/Util/ssBVH/ssBVH.cs
+    ///
+    /// Needs all the scene's shapes, plus the indices of the shapes that were updated.
+    ///
+    pub fn optimize<Shape: BHShape>(&mut self,
+                                    refit_shape_indices: &HashSet<usize>,
+                                    shapes: &[Shape]) {
+        let mut refit_node_indices: HashSet<usize> = refit_shape_indices.iter()
+            .map(|x| shapes[*x].bh_node_index())
+            .collect();
+
+        // As long as we have refit nodes left, take the list of refit nodes
+        // with the highest depth (sweep nodes) and try to rotate them all
+        while refit_node_indices.len() > 0 {
+            let mut max_depth = 0;
+            let mut sweep_node_indices: Vec<usize> = Vec::new();
+
+            // Find max_depth and sweep_node_indices in one iteration
+            for refit_node_index in refit_node_indices.iter() {
+                let depth = match self.nodes[*refit_node_index] {
+                    BVHNode::Node { depth, .. } => depth,
+                    BVHNode::Leaf { depth, .. } => depth,
+                };
+
+                if depth > max_depth {
+                    max_depth = depth;
+                    sweep_node_indices.clear();
+                    sweep_node_indices.push(*refit_node_index);
+                } else if depth == max_depth {
+                    sweep_node_indices.push(*refit_node_index);
+                }
+            }
+
+            // Try to find a useful tree rotation with all nodes previously found
+            for sweep_node_index in sweep_node_indices.iter() {
+                // This node does not need to be checked again
+                refit_node_indices.remove(&sweep_node_index);
+
+                // TODO There might be multithreading potential here
+                let new_refit_node_index = self.try_rotate(*sweep_node_index, shapes);
+
+                // Instead of finding a useful tree rotation, we found another node
+                // that we should check, so we add its index to the refit_node_indices.
+                if let Some(index) = new_refit_node_index {
+                    refit_node_indices.insert(index);
+                }
+            }
+        }
+    }
+
+    /// Checks if there is a way to rotate a child and a grandchild (or two grandchildren) of
+    /// the given node (specified by `node_index`) that would improve the `BVH`.
+    /// If there is, the best rotation found is performed.
+    /// Relies on the children and transitive children of the given node having correct AABBs.
+    ///
+    /// Returns Some(index_of_node) if a new node was found that should be used for optimization.
+    ///
+    fn try_rotate<Shape: BHShape>(&mut self, node_index: usize, shapes: &[Shape]) -> Option<usize> {
```

Comment:
I don't see any way to do that so it brings an actual benefit. Any splits into additional scopes would just yield more code around it to keep the borrow checker happy.
I've been generously commenting its sections instead.